geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/Frequency.scala

```diff
@@ -0,0 +1,212 @@
+import com.vividsolutions.jts.geom.Geometry
+import org.locationtech.geomesa.curve.Z2SFC
+import org.locationtech.sfcurve.IndexRange
+import org.opengis.feature.simple.SimpleFeature
+
+import scala.reflect.ClassTag
+
+/**
+  *  Estimates frequency counts at scale
+  *
+  * @param attribute attribute index for the attribute the sketch is being made for
+  * @param eps (epsilon) with probability at least @see confidence, estimates will be within eps * N
+  * @param confidence percent - with probability at least confidence, estimates will be within @see eps * N
+  * @param precision for geometry types, this is the number of bits of z-index to keep (max of 64)
+  *                  (note: first 2 bits do not hold any info)
+  *                  for date types, this is the number of milliseconds to group for binning
+  *                  for number types, this is the number of digits that will be grouped together
+  *                  for floating point types, this is the number of decimal places that will be considered
+  *                  for string types, this is the number of characters that will be considered
+  * @param ct class tag
+  * @tparam T type parameter, should match the type binding of the attribute
+  */
+class Frequency[T](val attribute: Int,
+                   val precision: Int,
+                   val eps: Double = 0.005,
+                   val confidence: Double = 0.95)(implicit ct: ClassTag[T]) extends Stat {
+
+  override type S = Frequency[T]
+
+  private [stats] val sketch: CountMinSketch = new CountMinSketch(eps, confidence, Frequency.Seed)
+
+  private val addAttribute = Frequency.add[T](ct.runtimeClass.asInstanceOf[Class[T]], precision)
+  private val getCount = Frequency.count[T](ct.runtimeClass.asInstanceOf[Class[T]], precision)
+
+  /**
+    * Gets the count for a given value
+    *
+    * @param value value to consider
+    * @return count of the value
+    */
+  def count(value: T): Long = getCount(sketch, value)
+
+  /**
+    * Gets the count for a given value, which has already been converted into a string. Useful
+    * if you know the string key space ahead of time.
+    *
+    * @param value value to consider, converted into an appropriate string key
+    * @return count of the value
+    */
+  def countDirect(value: String): Long = sketch.estimateCount(value)
+
+  /**
+    * Gets the count for a given value, which has already been converted into a long. Useful
+    * if you know the long key space ahead of time (e.g. with z-values).
+    *
+    * @param value value to consider, converted into an appropriate long key
+    * @return count of the value
+    */
+  def countDirect(value: Long): Long = sketch.estimateCount(value)
+
+  /**
+    * Number of observations in the frequency map
+    *
+    * @return number of observations
+    */
+  def size: Long = sketch.size()
+
+  override def observe(sf: SimpleFeature): Unit = {
+    val value = sf.getAttribute(attribute)
+    if (value != null) {
```

Comment:
Do we really want to discard all information about nulls?  Wouldn't it be more useful to maintain this as a separate count?  I understand that the sketch should only use the non-null entries, but knowing how many nulls seems as if it could also be extremely useful.  Just a thought.